### ISSUE TITLE
AGS4: 'Castle Escape Chapter 2' is said to crash in Linux but not in Windows

### ISSUE THREAD

**Issue by fernewelten** [adventuregamestudio/ags#1604]
**Describe the bug**
I've been told that my most recent game 'Castle Escape Chapter 2' will crash on Linux, but not on Windows, shortly after the end of the starting cutscene. I can't test this because I don't own a Linux box.
 
**AGS Version**
In the AGS4 branch. It's been a little while, and I'm no longer sure where exactly: __All__ the 4.0 versions identify as “AGS-3.99.99.0.exe”.
Might be commit 9db18d5

**Game**
Windows version: https://www.loquimur-online.de/ags/MAGS/CastleEscapeChapter2/Windows.zip
Linux version: https://www.loquimur-online.de/ags/MAGS/CastleEscapeChapter2/Linux.7z

**To Reproduce**
As sthomannch writes here: https://www.adventuregamestudio.co.uk/forums/index.php?topic=59811.msg636644094#msg636644094
“After the intro, when the mouse pointer changes to an arrow so that you can interact, the game crashes after about 1 second with a segmentation fault.”
```
...
Setting up window
Initialize sprites
Engine initialization complete
Starting game
Segmentation fault (core dumped)
```
Note: “**in fact, no coredump around.**”

**Expected behaviour**
Obviously, that the game doesn't segfault.

**Screenshots**
n/a

**Desktop (please complete the following information):**
Linux

**Smartphone (please complete the following information):**
n/a

**Additional context**
n/a

**Comment by rofl0r** [comment]
the "not owning a linux box" part can be solved easily, as 99% of linux distros are downloadable at no cost and can be readily installed in a virtual machine (oracle's virtualbox can be used for that, it's free and opensource too). the only cost are ~4 gb of temporary hdd space.

**Comment by ivan-mogilko** [comment]
According to gdb, the segfault occurs at Sc_DynamicArray_Length, line 180 (this is basically a call to API_SCALL_PINT_OBJ .. etc)

**Comment by rofl0r** [comment]
> I don't know how accurate gdb's information is (don't have much experience with it, so could be using or setting it up wrong), but according to it the received pointer value of "params" looks bad (0x7fffff.. or something).

it's usually very accurate. what's the complete value of that pointer ? if it's using the full 64bit space, it's probably due to a truncated (64-to-32) and sign-extended pointer, which usually happens due to "implicit function declarations", which should be visible as a warning. you can also build with -Werror=implicit-function-declaration to find whether that's what's happening.

**Comment by ericoporto** [comment]
@ivan-mogilko it's possible to reproduce it in Windows when building it for 64-bit.
<details><img src="https://user-images.githubusercontent.com/2244442/163504761-57d942da-208f-4a01-90a1-3e79955ad491.png"></details>

**Comment by ivan-mogilko** [comment]
The code there actually seem wrong, I am not certain why it worked in 32-bit either.

**Comment by fernewelten** [comment]
I think it _is_ wrong, starting at line 173
```size_t const k_offset_to_num_elements = -2u;```

The term `-2u` is obviously an underflow. I don't know why I did that this way, must have been a brain fart. The calculation ought to be done in `int`. AFAIK, it's okay to add an `int` to a pointer even if the `int` is negative.

```
    int const k_offset_to_num_elements = -2u;
    auto typed_dynarray = static_cast<int32_t *>(untyped_dynarray);
    return typed_dynarray[k_offset_to_num_elements] & ~ARRAY_MANAGED_TYPE_FLAG;
```

**Comment by ivan-mogilko** [comment]
Fixed by #1614 (tested on 64-bit Ubuntu vm).
